### PR TITLE
Yarn Berry: Switch from from node_modules to PnP 

### PR DESCRIPTION
**Pull request by kachkaev**
Continues #160, replaces #29

---

Key steps:

1.  Edit `.yarnrc.yml`: 

    ```diff
    -nodeLinker: node-modules
    +nodeLinker: pnp
    ```
1.  Run 

    ```sh
    rm -rf node_modules
    rm -rf cli/node_modules ## specific to this project
    ```

1.  Update `.gitignore` and other `*ignore` files:

    ```diff
    -node_modules
    +.pnp.*
    ```
    It might be worth keeping `node_modules` for some time if you expect to switch between PnP and non-PnP branches.

1.  Run   
    ```ts
    yarn install
    yarn dlx @yarnpkg/sdks vscode
    ```

1.  Remove `nodeLinker: pnp` from `.yarnrc.yml` (doing this in step 1 breaks `yarn install` in step 2).

1.  Add `require("@rushstack/eslint-patch/modern-module-resolution");` and `parserOptions: { tsconfigRootDir: __dirname },` to `.eslintrc.js` – see https://www.npmjs.com/package/@rushstack/eslint-patch. Thanks to @merceyz for [the tip](https://github.com/kachkaev/njt/pull/161#issuecomment-974886486)!

1. Track https://github.com/microsoft/vscode/issues/108866 and don’t use multi-root workspaces in VSCode until resolved
   - **Workaround:** one VSCode window → only one project folder 🙁

1.  Track https://github.com/vercel/vercel/issues/5280 for a permanent fix of Next.js deployment to Vercel

    - From Vercel’s knowledge base: https://vercel.com/knowledge/does-vercel-support-yarn-2

        > Serverless Functions and SSR (Server-side Rendering) will not work with Yarn 2.  

    - Case study: https://github.com/yarnpkg/berry/issues/2031
    
    - **[Workaround](https://github.com/vercel/vercel/issues/5280#issuecomment-744834737)**: Add `FORCE_BUILDER_TAG=canary` in Vercel dashboard

### CODE REVIEW

**Comment by merceyz** [comment]
> Hoist ESLint plugins

You can add https://github.com/microsoft/rushstack/tree/9c81203c95dcfab1631e9fd12913c7a3681d7360/stack/eslint-patch to your ESLint config to fix their resolution instead

File: package.json [hunk 1]
```diff
@@ -28,19 +28,11 @@
...
-    "eslint-plugin-jest": "^25.2.4",
-    "eslint-plugin-json": "^3.1.0",
-    "eslint-plugin-testing-library": "^5.0.0"
-  },
   "dependencies": {
     "@kachkaev/react-inline-js": "^0.2.0",
     "hosted-git-info": "^4.0.2",
     "lru-cache": "^6.0.0",
-    "next": "^12.0.5-canary.6",
+    "next": "12.0.4",
```
Comment (kachkaev) [inline package.json:35]: https://github.com/vercel/next.js/issues/31552

File: package.json [hunk 2]
```diff
@@ -55,15 +47,25 @@
...
+    "eslint-plugin-simple-import-sort": "^5.0.3",
+    "eslint-plugin-testing-library": "^5.0.0",
     "husky": "^7.0.4",
     "lint-staged": "^12.0.3",
     "markdownlint-cli": "^0.30.0",
     "ncp": "^2.0.0",
     "npm-run-all": "^4.1.5",
     "prettier": "~2.2.1",
-    "typescript": "^4.4.4"
+    "typescript": "4.4.4"
```
Comment (kachkaev) [inline package.json:68]: https://github.com/microsoft/TypeScript/issues/46874

File: .yarnrc.yml [hunk 3]
```diff
@@ -1,2 +1,9 @@
-nodeLinker: node-modules
+packageExtensions:
+  eslint-module-utils@*:
+    dependencies:
+      eslint-import-resolver-node: "*"
+  styled-components@*:
+    dependencies:
+      react-is: "*"
```
Comment (kachkaev) [inline .yarnrc.yml:7]: https://github.com/styled-components/styled-components/issues/3082

File: .yarnrc.yml [hunk 4]
```diff
@@ -1,2 +1,9 @@
-nodeLinker: node-modules
+packageExtensions:
+  eslint-module-utils@*:
+    dependencies:
+      eslint-import-resolver-node: "*"
```
Comment (kachkaev) [inline .yarnrc.yml:4]: https://github.com/yarnpkg/berry/issues/1427#issuecomment-637538879

**Comment by kachkaev** [comment]
Great, thanks for your tip @merceyz! Adding `require("@rushstack/eslint-patch/modern-module-resolution");` to `.eslintrc` worked like a charm! I updated the migration instruction in the first comment and mentioned you there.

If you have any other tips, please fell free to share! 🙌

File: package.json (hunk 2 above)
Comment (merceyz) [reply package.json:68]: See https://github.com/yarnpkg/berry/issues/3722#issuecomment-974634888

Comment (kachkaev) [reply package.json]: Nice, thanks! Running `yarn set version canary && yarn install` allowed me to use TS 4.5.2 with PnP 🎉

I’m now getting a bunch of `--experimental-loader` warnings, but these can be looked into separately.

Screenshot 2021-11-21 at 21 21 10

Comment (kachkaev) [reply package.json]: Asked about the warnings here: https://github.com/yarnpkg/berry/issues/3722#issuecomment-974900819

Comment (merceyz) [reply package.json]: Answered in https://github.com/yarnpkg/berry/issues/3722#issuecomment-974908156

File: package.json (hunk 1 above)
Comment (kachkaev) [reply package.json]: https://github.com/vercel/next.js/issues/31812

**Comment by rauchg** [comment]
I see you're using Vercel! What are things we could do to support PnP better on our platform, if any? Thanks in advance.

**Comment by kachkaev** [comment]
👋 @rauchg! Unfortunately Vercel does not currently support Yarn PnP. Here are some links I could dig out while working on this PR:

- https://github.com/vercel/vercel/issues/5280.

- From Vercel’s knowledge base: https://vercel.com/knowledge/does-vercel-support-yarn-2

        > Serverless Functions and SSR (Server-side Rendering) will not work with Yarn 2.  

- Case study: https://github.com/yarnpkg/berry/issues/2031
    
As a workaround proposed in https://github.com/vercel/vercel/issues/5280#issuecomment-744834737, I had to set `FORCE_BUILDER_TAG=canary` in the Vercel dashboard. It works, but I guess what it does is setting `YARN_NODE_LINKER=node-modules` to turn off Yarn PnP.

I hope PnP will get some priority this year because Yarn installs are significantly faster without `node_modules`. For you at Vercel this means happier customers and smaller infrastructure bills 😁

Looking at the state of my experiment (this PR), I’m somewhat sure that PnP will start setting off in 2022. Tooling support was way worse a year ago when I started, but it’s pretty much there now!

File: package.json (hunk 1 above)
Comment (kachkaev) [reply package.json]: Fixed in `12.0.8-canary.21`

**Comment by kachkaev** [comment]
Mission complete! 🕺

https://twitter.com/kachkaev/status/1488630228842917891

**Pull request merged**